### ISSUE TITLE
Compilation issues with rust starting 1.81 

### ISSUE THREAD

**Issue by yanshay** [sammhicks/picoserve#57]
I upgraded from rust 1.80 to 1.82 and previous picoserve code that worked (based on embassy example here) is not compiling:

```
type AppRouter = impl picoserve::routing::PathRouter;

#[embassy_executor::task(pool_size = WEB_TASK_POOL_SIZE)]
async fn web_task(
    task_id: usize,
    stack: &'static Stack<WifiDevice<'static, WifiStaDevice>>,
    app: &'static picoserve::Router<AppRouter>,
    config: &'static picoserve::Config<Duration>,
) {

}
```
I get the following compilation errors on the above piece of code:
```
error: item does not constrain `AppRouter::{opaque#0}`, but has it in its signature
  --> src/framework/web_config.rs:30:10
   |
30 | async fn web_task(
   |          ^^^^^^^^
   |
   = note: consider moving the opaque type's declaration and defining uses into a separate module
note: this opaque type is in the signature
  --> src/framework/web_config.rs:25:18
   |
25 | type AppRouter = impl picoserve::routing::PathRouter;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: item does not constrain `AppRouter::{opaque#0}`, but has it in its signature
  --> src/framework/web_config.rs:36:3
   |
36 |   ) {
   |  ___^
37 | |
38 | | }
   | |_^
   |
   = note: consider moving the opaque type's declaration and defining uses into a separate module
note: this opaque type is in the signature
  --> src/framework/web_config.rs:25:18
   |
25 | type AppRouter = impl picoserve::routing::PathRouter;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: item does not constrain `AppRouter::{opaque#0}`, but has it in its signature
  --> src/framework/web_config.rs:29:1
   |
29 | #[embassy_executor::task(pool_size = WEB_TASK_POOL_SIZE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: consider moving the opaque type's declaration and defining uses into a separate module
note: this opaque type is in the signature
  --> src/framework/web_config.rs:25:18
   |
25 | type AppRouter = impl picoserve::routing::PathRouter;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `embassy_executor::task` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Any ideas how to solve this?

**P.S.**
Originally I thought it had to do with the need to move away from `make_static` which is no longer working in 1.81 and above, but after some more tests, it looks to me like the root cause is the question above. 
Below is the previous question that I think is no longer relevant, but keeping it here in case it is.

---

In the embassy examples the code make use of `make_static`.

e.g.
    ```let app = make_static!(make_app());```

afaik, `make_static` no longer work starting rust 1.81 (https://github.com/embassy-rs/static-cell/issues/16)

- Are you getting the examples to compile using rust 1.81 and above? 

So, instead, I've seen the use of `mk_static` macro which requires supplying also the type:

```
macro_rules! mk_static {
    ($t:ty,$val:expr) => {{
        static STATIC_CELL: static_cell::StaticCell<$t> = static_cell::StaticCell::new();
        #[deny(unused_attributes)]
        let x = STATIC_CELL.uninit().write(($val));
        x
    }};
}
```

When I tried to use it as follows:
```
mk_static!(Router<AppRouter>, make_app());
```

I get an error:

```
error: item does not constrain `AppRouter::{opaque#0}`, but has it in its signature
   --> src/main.rs:48:16
    |
48  |         static STATIC_CELL: static_cell::StaticCell<$t> = static_cell::StaticCell::new();
    |                ^^^^^^^^^^^
    |
   ::: src/framework/web_config.rs:148:15
    |
148 |     let app = mk_static!(Router<AppRouter>, make_app());
    |               --------------------------------------------------------------------- in this macro invocation
    |
    = note: consider moving the opaque type's declaration and defining uses into a separate module
note: this opaque type is in the signature
   --> src/framework/web_config.rs:25:18
    |
25  | type AppRouter = impl picoserve::routing::PathRouter;
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `mk_static` (in Nightly builds, run with -Z macro-backtrace for more info)

```

I also tried using `mk_static!(Router<Impl PathRouter>` and got a different error.

I also tried using `Rc` to pass the `make_app()` result to the `web_task` and couldn't make it work.

So practically, once I switched to the new rust version I can't get it to compile with an async app and several web_tasks.

I'm no expert, but it looks like now (with new rustc) it's required to know the concrete type for this mk_static, and the concrete type in picoserve seems to be dynamic based on the exact app and can't be deduced manually. The compiler used to be able to infer that with make_static and that solved the issues. But again, I could be totally wrong.

Any ideas/guidelines how to resolve this?

**Comment by lu-zero** [comment]
So far the workaround commonly used (even in the picoserve repo) is pinning the last known working nightly compiler with a rust-toolchain.toml

**Comment by yanshay** [comment]
> So far the workaround commonly used (even in the picoserve repo) is pinning the last known working nightly compiler with a rust-toolchain.toml

It's a workaround if you don't need features from new compilers. There are cases where the the latest compiler version is required.

**Comment by lu-zero** [comment]
in that case please look in rust-lang for an issue related to it and link this one ^^;

**Comment by yanshay** [comment]
What do you mean by issue in `rust-lang`?

Here is my understanding so far, maybe it will help:
- To use a pool of picoserve web_tasks to serve requests with embassy it's required to store the picoserve app in some static (or Rc if alloc available). For that `make_static!` was used.
- `make_static!` is no longer work with newer rust versions starting 1.81. That's explained in https://github.com/embassy-rs/static-cell/issues/16. (I've been told this is due to reliance on nightly TAIT where things changed. But that's beyond my understanding - https://github.com/esp-rs/esp-hal/discussions/2577#discussioncomment-11329214)
- The available alternatives requires requires providing the concrete type explicitly and not have the compiler infer it. This worked for me in all cases, and sometimes telling the type is difficult but there are ways to get the compiler to hint in the error message as to what type it expects (https://github.com/esp-rs/esp-hal/discussions/2577#discussioncomment-11328959)
- When I did it with picoserve make_app result (which goes into the static), I got a very complex type that changed as I changed the routes. I don't know how that type is generated, or maybe the compiler hint is wrong. Anyway at least I couldn't get it to work. I tried both static and Rc for sharing the app among the web_tasks.

So eventually I had to move to a single web_task. For my current app I can make it work, but it practically make picoserve unusable for more than the most basic application.

I could be totally wrong, just sharing my current understanding.

**Comment by lu-zero** [comment]
The problem is that rustc has a bug in dealing with the opaque types and if you try to follow what the diagnostic suggest it errors out with an equally misleading error...

**Comment by yanshay** [comment]
> The problem is that rustc has a bug in dealing with the opaque types and if you try to follow what the diagnostic suggest it errors out with an equally misleading error...

So is there a solution to get a picoserve app working with multiple web_task?
Or isn't there a solution until that bug is resolved?

**Comment by sammhicks** [comment]
I think that I might have a solution, which I'm demonstrating in the [`from_tait_to_itiat`](https://github.com/sammhicks/picoserve/tree/from_tait_to_itiat) branch.

Please let me know if the solution doesn't work for you :)

**Comment by yanshay** [comment]
> I think that I might have a solution, which I'm demonstrating in the [`from_tait_to_itiat`](https://github.com/sammhicks/picoserve/tree/from_tait_to_itiat) branch.
> 
> Please let me know if the solution doesn't work for you :)

Solution works, thanks!

Just one comment:
My `build_app` requires parameters, and the new `build_app` on `AppBuilder`, which comes from lib.rs, doesn't support it, so I just used my own `MyAppBuilder` and `MyRouter` and added the params to `build_app` and could use it with the main release. So IMO it would even be better to leave everything in the examples only and not add anything to the crate so not to confuse and see that as a crate struct that has to be used.

**Comment by sammhicks** [comment]
@yanshay 

> My `build_app` requires parameters, and the new `build_app` on `AppBuilder`, which comes from lib.rs, doesn't support it

`AppBuilder::build_app` takes `self`, so you should be able to pass data by having the struct which implements `AppBuilder` to have fields, as follows.

For example

```rust
struct AppProps {
    message: &'static str,
}

impl AppBuilder for AppProps {
    type Router = picoserve::Router<impl picoserve::routing::PathRouter>;

    fn build_app(self) -> Self::Router {
        let Self { message } = self;

        picoserve::Router::new().route("/", get(move || async move { message }))
    }
}
```

**Comment by sammhicks** [comment]
Resolved by 00266199e6db138fe61e64034671458ab2f4135d
